core/src/main/java/io/methvin/watchservice/MacOSXListeningWatchService.java

```diff
@@ -145,6 +147,8 @@ public AbstractWatchKey register(
             flags);
 
     final CFRunLoopThread thread = new CFRunLoopThread(streamRef, file.toFile());
+    callback.setRunLoopThread(thread);
```

Comment:
Yeah, so I think a simpler solution is something like:
```java
callback.onClose(() -> close(thread, callback, path));
```
since the only thing you need is a way for the callback to signal to the watcher that it is done, and the watcher is responsible for cleaning up all the state.

Reply:
I cannot see an callback.onClose method signature, it only has invoke.  In the callback I check if the paths are gone and close down the thread, so it needs access to the thread.

Reply:
Sorry, I don't think I was clear here. What I meant was to create an additional method on the `MacOSXListeningCallback` like this:

```java
public void onClose(Callable<Void> callback) {
  this.onCloseCallback = callback;
}
```

then the `MacOSXListeningCallback` can call `onCloseCallback.call()` rather than calling `watchKey.watchService().close(runLoopThread, this, Paths.get(absPath))`. This way the callback doesn't need to have access to the `runLoopThread` or the `watchService`.

What I was trying to do here is keep the actual callback logic as simple as possible, since I think passing around extra dependencies made it more confusing. If you can suggest a better way let me know.

Reply:
done